File: CoinChange2Exhaustive.java
// Time Complexity : O(2 pow (m + n))
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No
class Solution {
    public int change(int amount, int[] coins) {
        //null case
        if(coins.length == 0 || amount == 0)
            return 0;
        return helper(amount,coins,0,0);
    }
    
    private int helper(int amount, int[] coins, int index, int result){
        //base case
        if(result == amount)
            return 1;
        if(result > amount || index > coins.length -1)
            return 0;
        //chose the coin
        int case1 = helper(amount,coins,index,result + coins[index]);
        //do not chose the coin
        int case2 = helper(amount,coins,index + 1,result);
        
        return case1 + case2;
    }
}

File: coin-change-2-exhaustive.cpp
// Time Complexity : O(2^(coins.length + amount))
// Space Complexity : O(amount)
// Did this code successfully run on Leetcode : No (TLE)

#include <iostream>
#include <vector>

using namespace std;

class Solution {
public:
    int change(int amount, vector<int>& coins) {
        return helper(amount, 0, coins);
    }
    int helper(int amount, int index, vector<int>& coins) {
        if (amount < 0)
            return 0;
        if (amount == 0)
            return 1;
        if (index >= coins.size())
            return 0;

        int count = helper(amount, index + 1, coins) + helper(amount - coins[index], index, coins);
        return count;
    }
};